feat: FO-97 | Add workspace-python-ds-pack-v1 image

saurabh0402 opened the pull request:
## Description
This PR adds the `workspace-python-ds-pack-v1` image. This image is just a copy of the existing [workspace-python](https://github.com/peak-platform/jupyter-image) image with a few changes:
- Fix for the broken build.
- Removes a few not needed packages.

## Review Checks
- [x] 📝 The commit message is clear and descriptive
- [x] 🔐 The Security Considerations section in the PR description is complete - **Please do not remove this**
- [ ] ✅ Tests for the changes have been added and run successfully including the new changes
- [x] 📄 Documentation has been added / updated (for bug fixes / features)


## Dependencies
None

**Does this PR introduce a breaking change?** 
- [ ] Yes
- [x] No

## Security Considerations
- No Peak-specific data has been added to the image.
- No secrets have been added to the image.
- Since, users don't have root access in workspaces we don't need to create a non-root user.

## Types of change
- [ ] 🐛 Bugfix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 🛠 Adding or updating configuration files, development scripts etc.
- [ ] ♻️ Refactoring (no functional changes, no API changes)
- [ ] 🧹 Chore (removing redundant files, fixing typos, etc.)
- [ ] 📄 Documentation Update 
- [ ] ❓ Other (if none of the other choices apply)

## Testing
- Try building and running the image locally. It should work fine.
- If you have access to the Peak platform, you can build the image and use it to create a workspace.

peak-jiteshahuja commented on workspace/python/ds-pack/v1/README.md:
```diff
@@ -0,0 +1,74 @@
+
+## Using the Image
+- The image can be directly used by using it in the workspace form.
+- If you need to install a few more dependencies, or add some use case-specific environment variables into the image, the image can easily to extended.
```
```suggestion
- If you need to install a few more dependencies, or add some use case-specific environment variables into the image, the image can easily be extended.
```

saurabh0402 replied:
Done.

peak-jiteshahuja reviewed (approved):
Looks goos and works fine.

conor-james-sheehan commented:
Hey can I ask why the data science Python packages have been installed in the root of the image?

The philosophy previously was to only install python packages required for running the jupyter workspaces (`jupyterhub` etc), and leave any other packages to the user to install in a virtual environment. This has a few advantages:
1. It doesn't bloat the image with packages the user might not need
2. It encourages best practices (installing packages in virtual environments, rather than in the root interpreter)
3. It reduces the maintenance burden of the image code - now there's a whole bunch more packages whose versions will need regularly updating e.g. when there are security bugfixes

lukedyer-peak commented on workspace/python/ds-pack/v1/requirements.txt:
```diff
@@ -0,0 +1,27 @@
+python-lsp-server[all]==1.4.1
+scikit-learn==1.1.1
+snowflake-connector-python[pandas]==2.7.7
+snowflake==0.0.3
```
This is a unknown [package](https://pypi.org/project/snowflake/) - potentially security risk

lukedyer-peak commented on workspace/python/ds-pack/v1/requirements.txt:
```diff
@@ -0,0 +1,27 @@
+pyls-mypy==0.1.8
+python-lsp-black==1.0.0
+python-lsp-server[all]==1.4.1
+scikit-learn==1.1.1
+snowflake-connector-python[pandas]==2.7.7
```
shouldn't be installed in the base image

lukedyer-peak commented on workspace/python/ds-pack/v1/requirements.txt:
```diff
@@ -0,0 +1,27 @@
+mypy-ls==0.5.1
+nbconvert==6.2.0
+notebook==6.4.3
+pandas==1.4.2
+psycopg2==2.9.3
```
shouldn't be installed in the base image

lukedyer-peak commented on workspace/python/ds-pack/v1/requirements.txt:
```diff
@@ -0,0 +1,27 @@
+scikit-learn==1.1.1
+snowflake-connector-python[pandas]==2.7.7
+snowflake==0.0.3
+SQLAlchemy==1.4.36
```
shouldn't be installed in the base image

lukedyer-peak commented on workspace/python/ds-pack/v1/projector-proxy/jupyter_projector_proxy/__init__.py:
```diff
@@ -0,0 +1,41 @@
+            ),
+        },
+        # default projector config runs on port 9999
+        "port": 9999,
```
Does this work with multiple users at the same time.  We previously had a bug where a user would hog the port.  @conor-james-sheehan didn't you introduce changes to fix that?

conor-james-sheehan replied:
Nope - started but never found a solution, dropped it after a while

lukedyer-peak commented on workspace/python/ds-pack/v1/README.md:
```diff
@@ -0,0 +1,74 @@
+vim            8.1
+fish           3.0.2
+zsh            5.7.1
+R              3.5.2
```
R 4.x should be installed for parity with the R image

lukedyer-peak reviewed (changes requested):
See more comments in [slack thread](https://peak-ai.slack.com/archives/C039LM7B9GQ/p1655196921818279?thread_ts=1654755919.886399&cid=C039LM7B9GQ)

vanessavirgo94 commented:
Yeah I agree with @lukedyer-peak and @conor-james-sheehan. Makes sense what they are saying, I guess Workspaces are different that Workflows where some of the packages don't need to be in the Workspace image

saurabh0402 commented:
Hey @conor-james-sheehan @lukedyer-peak @vanessavirgo94, I have removed all the extra packages and updated R to 4.2.0. Can you please take a look now?

conor-james-sheehan reviewed (approved):
Thanks @saurabh0402 👍
